pkg/crcontrollers/storagegroup/storagegroupcontroller.go

```diff
@@ -0,0 +1,620 @@
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Current manual sg labeling support
+	log.Debugf("Handle manual change of storage group label of drive %s", drive.Name)
+
+	switch {
+	// add new storagegroup label to drive
+	case !acSGLabeled && driveSGLabeled:
+		return c.handleManualDriveStorageGroupLabelAddition(ctx, log, drive, ac, driveSGLabel, lvg != nil)
+
+	// remove storagegroup label from drive
+	case acSGLabeled && !driveSGLabeled:
+		volumes, err := c.crHelper.GetVolumesByLocation(ctx, drive.Spec.UUID)
+		if err != nil {
+			log.Errorf("Error when getting volumes on drive %s: %s", drive.Name, err.Error())
+			return ctrl.Result{Requeue: true}, err
+		}
+		if len(volumes) > 0 {
+			log.Warnf("We can't remove storage group %s label from drive %s with existing volumes",
+				acSGLabel, drive.Name)
+			if err := c.addDriveStorageGroupLabel(ctx, log, drive, acSGLabel); err != nil {
+				return ctrl.Result{Requeue: true}, err
+			}
+			return ctrl.Result{}, nil
+		}
+
+		sg := &sgcrd.StorageGroup{}
+		err = c.client.ReadCR(ctx, acSGLabel, "", sg)
+		switch {
+		case err == nil && c.isDriveSelectedByValidMatchFields(log, &drive.Spec, &sg.Spec.DriveSelector.MatchFields):
+			log.Warnf("We can't remove storage group %s label from drive %s still selected by this storage group",
+				acSGLabel, drive.Name)
+			if err := c.addDriveStorageGroupLabel(ctx, log, drive, acSGLabel); err != nil {
+				return ctrl.Result{Requeue: true}, err
+			}
+			return ctrl.Result{}, nil
+		case err != nil && !k8serrors.IsNotFound(err):
+			log.Errorf("Failed to read StorageGroup %s with error %v", acSGLabel, err)
+			return ctrl.Result{Requeue: true}, err
+
+		// the case that the storage-group label removal is valid and we should sync the removal to AC
+		default:
+			log.Debugf("Also remove the storage-group %s label of AC %s corresponding to drive %s", acSGLabel,
+				ac.Name, drive.Name)
+			if err := c.removeACStorageGroupLabel(ctx, log, ac); err != nil {
+				return ctrl.Result{Requeue: true}, err
+			}
+		}
+
+		// TODO restore the update of storagegroup label of drive
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (c *Controller) handleStorageGroupDeletion(ctx context.Context, log *logrus.Entry,
+	sg *sgcrd.StorageGroup) (ctrl.Result, error) {
+	drivesList := &drivecrd.DriveList{}
+	if err := c.client.ReadList(ctx, drivesList); err != nil {
+		log.Errorf("failed to read drives list: %s", err.Error())
+		return ctrl.Result{Requeue: true}, err
+	}
+	removalNoError := true
+	for _, drive := range drivesList.Items {
+		drive := drive
+		if drive.Labels[apiV1.StorageGroupLabelKey] == sg.Name {
+			if err := c.removeDriveAndACStorageGroupLabel(ctx, log, &drive, sg); err != nil {
+				log.Errorf("error in remove storage-group label from drive %s", err.Error())
+				removalNoError = false
+			}
+		}
+	}
+	if !removalNoError {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("error in removing storage-group label")
```

Comment:
Here, we have multiple drives and ACs for labeling removal. If we have multiple errors here, we cannot only return one of the multiple removal errs.

To avoid the err messages missing, I've just refined the logic to combine all the possible removal errs into one error to be returned if we encounter any error in the labeling removal of multiple drives and ACs.

Reply:
yes, I got you, but I mean replace removalNoError:=true with removalError:=false, and change !removalNoError(no no error) to removalError(err) maybe better.

Reply:
ok, your suggestion is better. Sorry for the misunderstanding...

But I just found that, with the introduction of labelRemovalErrMsgs to refine error messages for label removal, the bool flag removalNoError or removalErr can be deprecated... We can just use the length of labelRemovalErrMsgs to judge whether there is removalErr